File: android/app/src/main/java/com/noop/analytics/VitalityEngine.kt
package com.noop.analytics

import kotlin.math.abs

// VitalityEngine.kt — 0–100 "Vitality" wellness score + optional "Body Age in years".
// Byte-for-byte mirror of Strand/Packages/StrandAnalytics/Sources/StrandAnalytics/VitalityEngine.swift.
//
// INDEPENDENT implementation of the published method WHOOP's "Healthspan / WHOOP Age" also uses (NOT
// medical advice; a wellness comparison, never a clinical biological age): map each wearable input to its
// published all-cause-mortality hazard ratio vs a population reference, sum the log-hazards with an
// overlap correction (correlated inputs), and convert to a "years of aging" offset via the Gompertz
// mortality-rate doubling time (~8 years per doubling). Body Age = chronological age + Δage; average =
// own age, healthier = younger. Presented with a ±band and a hard "wellness, not a clinical age" rule.
object VitalityEngine {

    private const val lnHazardPerYear = 0.6931471805599453 / 8.0   // ln(2)/8 ≈ 0.0866
    private const val overlapShrink = 0.75
    const val minBodyAge = 20.0
    const val maxBodyAge = 90.0
    private const val vitalityPerYear = 2.5
    const val minFactors = 3
    const val bandYears = 5.0

    data class Inputs(
        val chronoAge: Double,
        val restingHR: Double? = null,
        val vo2max: Double? = null,
        val expectedVO2max: Double? = null,
        val sleepHours: Double? = null,
        val sleepConsistency: Double? = null,
        val rmssd: Double? = null,
        val rmssdNorm: Double? = null,
        val steps: Double? = null,
    )

    data class Contribution(val key: String, val label: String, val lnHazard: Double)

    data class Result(
        val vitality: Double,
        val bodyAge: Double,
        val chronoAge: Double,
        val deltaYears: Double,
        val bandYears: Double,
        val contributions: List<Contribution>,
        val factorsUsed: Int,
    )

    /** Per-factor signed log-hazard vs the population reference (positive ages you, negative protective).
     *  Conservative published per-unit hazard ratios — see the Swift file for citations. */
    fun contributions(inputs: Inputs): List<Contribution> {
        val out = ArrayList<Contribution>()
        inputs.restingHR?.let {
            out.add(Contribution("rhr", "Resting heart rate", ((it - 65) / 10) * 0.100))
        }
        val vo2 = inputs.vo2max; val exp = inputs.expectedVO2max
        if (vo2 != null && exp != null && exp > 0) {
            out.add(Contribution("vo2max", "Cardio fitness", ((exp - vo2) / 3.5).coerceIn(-4.0, 4.0) * 0.130))
        }
        inputs.sleepHours?.let {
            val dev = maxOf(0.0, abs(it - 7.5) - 0.5)
            out.add(Contribution("sleep", "Sleep duration", dev.coerceIn(0.0, 3.0) * 0.110))
        }
        inputs.sleepConsistency?.let {
            out.add(Contribution("consistency", "Sleep regularity", (0.75 - it.coerceIn(0.0, 1.0)) * 0.450))
        }
        val h = inputs.rmssd; val norm = inputs.rmssdNorm
        if (h != null && norm != null && norm > 0) {
            out.add(Contribution("hrv", "Heart-rate variability", ((norm - h) / norm).coerceIn(-1.0, 1.0) * 0.160))
        }
        inputs.steps?.let {
            val deficit = (7000 - it.coerceIn(0.0, 11000.0)) / 1000
            out.add(Contribution("steps", "Daily steps", deficit.coerceIn(-4.0, 4.0) * 0.064))
        }
        return out
    }

    /** Full Vitality + Body Age. Returns null until at least [minFactors] inputs are present. */
    fun compute(inputs: Inputs): Result? {
        if (inputs.chronoAge <= 0) return null
        val contribs = contributions(inputs)
        if (contribs.size < minFactors) return null
        val sumLn = contribs.sumOf { it.lnHazard } * overlapShrink
        val deltaAge = sumLn / lnHazardPerYear
        val bodyAge = (inputs.chronoAge + deltaAge).coerceIn(minBodyAge, maxBodyAge)
        val delta = inputs.chronoAge - bodyAge
        val vitality = (50 + delta * vitalityPerYear).coerceIn(0.0, 100.0)
        return Result(vitality, bodyAge, inputs.chronoAge, delta, bandYears, contribs, contribs.size)
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/VitalityEngine.swift
import Foundation

// VitalityEngine.swift — a transparent 0–100 "Vitality" wellness score + an optional "Body Age in years".
//
// INDEPENDENT implementation of the published, peer-reviewed method WHOOP's "Healthspan / WHOOP Age" also
// uses (NOT medical advice; a wellness comparison, never a clinical biological age): map each wearable-
// measurable input to its published ALL-CAUSE-MORTALITY hazard ratio relative to a population reference,
// sum the log-hazards with an overlap correction (the inputs are correlated, so the naive sum overstates),
// and convert that combined hazard into a "years of aging" offset using the Gompertz mortality-rate
// doubling time (mortality roughly doubles every ~8 years, so 1 doubling of hazard ≈ 8 years of age).
//
// Body Age = chronological age + Δage. An average-for-their-age person nets ~0 and reads at their own age;
// healthier-than-average reads younger, less healthy reads older. Presented with a ±band and a hard
// "wellness trend, not a biological/clinical age" disclaimer, gated on a minimum number of inputs.
//
// Per-factor hazard ratios are taken from large cohorts / meta-analyses (UK Biobank, FRIEND, pooled
// step- and activity-mortality meta-analyses, sleep-regularity and HRV cohorts). They are deliberately
// CONSERVATIVE and the model is clamped, because this is a wellness estimate, not a diagnosis.
public enum VitalityEngine {

    // Gompertz: mortality-rate doubling time ≈ 8 years → ln(hazard) per year of age = ln(2)/8.
    static let lnHazardPerYear = 0.6931471805599453 / 8.0   // ≈ 0.0866
    /// Correlated inputs (fitness, RHR, activity all move together) → shrink the naive log-hazard sum so
    /// we don't multiply the same underlying signal several times. 0.75 is a deliberately gentle shrink.
    static let overlapShrink = 0.75
    /// Body Age is clamped to a sane band; Vitality maps Δage linearly around 50 (= "at your age").
    static let minBodyAge = 20.0, maxBodyAge = 90.0
    static let vitalityPerYear = 2.5   // each year younger than your age = +2.5 Vitality points

    /// The wearable inputs Vitality reads. All optional — the score uses whatever is present (≥ minFactors).
    public struct Inputs: Equatable, Sendable {
        public var chronoAge: Double
        public var restingHR: Double?          // bpm
        public var vo2max: Double?             // ml/kg/min (e.g. from FitnessAgeEngine)
        public var expectedVO2max: Double?     // age/sex-expected ml/kg/min (the reference for vo2max)
        public var sleepHours: Double?         // mean nightly sleep
        public var sleepConsistency: Double?   // 0–1 regularity (1 = perfectly regular)
        public var rmssd: Double?              // ms, nocturnal HRV
        public var rmssdNorm: Double?          // age/sex-normative RMSSD (the reference)
        public var steps: Double?              // mean daily steps
        public init(chronoAge: Double, restingHR: Double? = nil, vo2max: Double? = nil,
                    expectedVO2max: Double? = nil, sleepHours: Double? = nil,
                    sleepConsistency: Double? = nil, rmssd: Double? = nil,
                    rmssdNorm: Double? = nil, steps: Double? = nil) {
            self.chronoAge = chronoAge; self.restingHR = restingHR; self.vo2max = vo2max
            self.expectedVO2max = expectedVO2max; self.sleepHours = sleepHours
            self.sleepConsistency = sleepConsistency; self.rmssd = rmssd
            self.rmssdNorm = rmssdNorm; self.steps = steps
        }
    }

    /// One factor's contribution: its label and signed log-hazard vs the population reference
    /// (positive = ages you, negative = protective).
    public struct Contribution: Equatable, Sendable {
        public let key: String
        public let label: String
        public let lnHazard: Double
        public init(key: String, label: String, lnHazard: Double) {
            self.key = key; self.label = label; self.lnHazard = lnHazard
        }
    }

    public struct Result: Equatable, Sendable {
        public let vitality: Double        // 0–100 (50 = typical for your age)
        public let bodyAge: Double         // years, clamped
        public let chronoAge: Double
        public let deltaYears: Double      // chronoAge − bodyAge (positive = younger than your age)
        public let bandYears: Double
        public let contributions: [Contribution]   // for the "what's driving this" breakdown
        public let factorsUsed: Int
        public init(vitality: Double, bodyAge: Double, chronoAge: Double, deltaYears: Double,
                    bandYears: Double, contributions: [Contribution], factorsUsed: Int) {
            self.vitality = vitality; self.bodyAge = bodyAge; self.chronoAge = chronoAge
            self.deltaYears = deltaYears; self.bandYears = bandYears
            self.contributions = contributions; self.factorsUsed = factorsUsed
        }
    }

    /// Minimum distinct factors before we'll show a number (honesty gate).
    public static let minFactors = 3
    public static let bandYears = 5.0

    private static func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double { min(hi, max(lo, v)) }

    /// Compute the per-factor log-hazard contributions present in `inputs`. Each references a population
    /// value, so an average person nets ~0. Published per-unit hazard ratios (conservative, clamped):
    ///   • Resting HR: +~10.5% all-cause mortality per +10 bpm (UK Biobank / meta-analyses). ref 65.
    ///   • VO₂max: ~14% per MET (3.5 ml/kg/min) vs the age/sex-expected value (FRIEND). fitter = protective.
    ///   • Sleep duration: U-shaped, optimum ~7.5 h; only deviation beyond ±0.5 h adds hazard (~12%/h).
    ///   • Sleep regularity: most-regular vs least ≈ HR 0.70 (UK Biobank SRI). ref 0.75 of the 0–1 range.
    ///   • HRV (RMSSD): ~16% per relative SD below the age norm (lower HRV = higher hazard).
    ///   • Steps: ~12% per 1,000 steps/day up to ~7k, diminishing to ~11k (pooled step-mortality meta).
    public static func contributions(_ inputs: Inputs) -> [Contribution] {
        var out: [Contribution] = []
        if let rhr = inputs.restingHR {
            out.append(Contribution(key: "rhr", label: "Resting heart rate",
                                    lnHazard: ((rhr - 65) / 10) * 0.100))
        }
        if let vo2 = inputs.vo2max, let exp = inputs.expectedVO2max, exp > 0 {
            // (expected − vo2): if fitter than expected this is negative → protective.
            out.append(Contribution(key: "vo2max", label: "Cardio fitness",
                                    lnHazard: clamp((exp - vo2) / 3.5, -4, 4) * 0.130))
        }
        if let sh = inputs.sleepHours {
            let dev = max(0, abs(sh - 7.5) - 0.5)   // only deviation > ±0.5 h is a risk; optimum is neutral
            out.append(Contribution(key: "sleep", label: "Sleep duration",
                                    lnHazard: clamp(dev, 0, 3) * 0.110))
        }
        if let c = inputs.sleepConsistency {
            out.append(Contribution(key: "consistency", label: "Sleep regularity",
                                    lnHazard: (0.75 - clamp(c, 0, 1)) * 0.450))
        }
        if let h = inputs.rmssd, let norm = inputs.rmssdNorm, norm > 0 {
            out.append(Contribution(key: "hrv", label: "Heart-rate variability",
                                    lnHazard: clamp((norm - h) / norm, -1, 1) * 0.160))
        }
        if let s = inputs.steps {
            // Below ~7k each −1,000 steps adds hazard; protection caps near 11k (diminishing returns).
            let deficit = (7000 - clamp(s, 0, 11000)) / 1000
            out.append(Contribution(key: "steps", label: "Daily steps",
                                    lnHazard: clamp(deficit, -4, 4) * 0.064))
        }
        return out
    }

    /// Full Vitality + Body Age. Returns nil until at least `minFactors` inputs are present.
    public static func compute(_ inputs: Inputs) -> Result? {
        guard inputs.chronoAge > 0 else { return nil }
        let contribs = contributions(inputs)
        guard contribs.count >= minFactors else { return nil }
        let sumLn = contribs.reduce(0) { $0 + $1.lnHazard } * overlapShrink
        let deltaAge = sumLn / lnHazardPerYear              // +ve = ages you
        let bodyAge = clamp(inputs.chronoAge + deltaAge, minBodyAge, maxBodyAge)
        let delta = inputs.chronoAge - bodyAge              // +ve = younger than your age
        let vitality = clamp(50 + delta * vitalityPerYear, 0, 100)
        return Result(vitality: vitality, bodyAge: bodyAge, chronoAge: inputs.chronoAge,
                      deltaYears: delta, bandYears: bandYears, contributions: contribs,
                      factorsUsed: contribs.count)
    }
}
